File: android/digia-engage/src/test/java/com/digia/engage/internal/analytics/AnalyticsServiceTest.kt
package com.digia.engage.internal.analytics

import com.digia.engage.AnalyticsConfig
import com.digia.engage.DigiaExperienceEvent
import com.digia.engage.InAppPayload
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.test.StandardTestDispatcher
import kotlinx.coroutines.test.TestScope
import kotlinx.coroutines.test.advanceTimeBy
import kotlinx.coroutines.test.advanceUntilIdle
import kotlinx.coroutines.test.runCurrent
import kotlinx.coroutines.test.runTest
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotEquals
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Test

@OptIn(ExperimentalCoroutinesApi::class)
class AnalyticsServiceTest {

    private val testDispatcher = StandardTestDispatcher()
    private val testScope = TestScope(testDispatcher)

    private lateinit var sharedStore: InMemoryKeyValueStore

    @Before
    fun setUp() {
        sharedStore = InMemoryKeyValueStore()
    }

    // ── Helpers ──────────────────────────────────────────────────────────────

    private fun createService(
        config: AnalyticsConfig = AnalyticsConfig(flushIntervalMs = 10_000L),
        sender: AnalyticsSender = FakeSender { successResponse() },
        store: InMemoryKeyValueStore = sharedStore,
    ): AnalyticsService = AnalyticsService(
        config = config,
        apiKey = "test-api-key",
        identity = AnalyticsIdentityManager(store),
        queue = AnalyticsQueue(store),
        sender = sender,
        staticContext = mapOf("sdk_version" to "1.0.0", "sdk_platform" to "android"),
        scope = testScope,
    )

    private fun buildPayload(campaignKey: String) = InAppPayload(
        id = campaignKey,
        content = mapOf(
            "campaign_id" to "example-campaign",
            "campaign_key" to campaignKey,
            "campaign_type" to "guide",
        ),
    )

    // ── Tests ─────────────────────────────────────────────────────────────────

    @Test
    fun `anonymous ID is generated and stable`() = testScope.runTest {
        val service = createService()
        val id1 = service.identity.anonymousId
        val id2 = service.identity.anonymousId
        assertTrue(id1.isNotEmpty())
        assertEquals(id1, id2)
    }

    @Test
    fun `setUserId persists and clearUserId resets session`() = testScope.runTest {
        val service = createService()

        service.setUserId("user-123")
        assertEquals("user-123", service.identity.userId)

        val sessionBefore = service.identity.sessionId
        service.clearUserId()

        assertNull(service.identity.userId)
        assertTrue(service.identity.sessionId.isNotEmpty())
        assertNotEquals(sessionBefore, service.identity.sessionId)
    }

    @Test
    fun `queue drops oldest events when capacity exceeded`() = testScope.runTest {
        val service = createService(
            config = AnalyticsConfig(queueMaxEvents = 3, flushBatchSize = 100, flushIntervalMs = 10_000L),
        )
        // capture is synchronous — no advanceUntilIdle needed
        for (i in 0..4) {
            service.capture(DigiaExperienceEvent.Impressed, buildPayload("event-$i"))
        }

        assertEquals(3, service.queue.size())
        val entries = service.queue.peek(10)
        // oldest two (event-0, event-1) were dropped; event-2, event-3, event-4 remain
        assertEquals("event-2", entries[0].payload["campaign_key"])
        assertEquals("event-3", entries[1].payload["campaign_key"])
        assertEquals("event-4", entries[2].payload["campaign_key"])
    }

    @Test
    fun `event payload has correct structure and identity fields`() = testScope.runTest {
        val service = createService()
        service.capture(DigiaExperienceEvent.Impressed, buildPayload("payload-1"))
        // capture is synchronous — event is immediately in queue

        val entries = service.queue.peek(1)
        assertTrue(entries.isNotEmpty())
        val event = entries.first().payload

        assertEquals("Digia Experience Viewed", event["event_name"])
        assertEquals("example-campaign", event["campaign_id"])
        assertEquals("payload-1", event["campaign_key"])
        assertEquals("guide", event["campaign_type"])
        assertTrue((event["event_id"] as? String)?.isNotEmpty() == true)
        assertTrue((event["occurred_at"] as? String)?.isNotEmpty() == true)
        assertTrue((event["anonymous_id"] as? String)?.isNotEmpty() == true)
        assertTrue((event["session_id"] as? String)?.isNotEmpty() == true)
        assertNull(event["user_id"]) // not set — must be absent

        @Suppress("UNCHECKED_CAST")
        val props = event["properties"] as? Map<String, Any?>
        assertNotNull(props)
        assertEquals("1.0.0", props?.get("sdk_version"))
        assertEquals("android", props?.get("sdk_platform"))
    }

    @Test
    fun `event names map correctly for all experience event types`() = testScope.runTest {
        val service = createService()
        val payload = buildPayload("test")

        service.capture(DigiaExperienceEvent.Impressed, payload)
        service.capture(DigiaExperienceEvent.Clicked("cta-btn"), payload)
        service.capture(DigiaExperienceEvent.Dismissed, payload)
        // capture is synchronous — all three are immediately in queue

        val entries = service.queue.peek(10)
        assertEquals(3, entries.size)
        assertEquals("Digia Experience Viewed", entries[0].payload["event_name"])
        assertEquals("Digia Experience Clicked", entries[1].payload["event_name"])
        assertEquals("Digia Experience Dismissed", entries[2].payload["event_name"])

        // element_id only present for Clicked
        @Suppress("UNCHECKED_CAST")
        val clickProps = entries[1].payload["properties"] as? Map<String, Any?>
        assertEquals("cta-btn", clickProps?.get("element_id"))
        @Suppress("UNCHECKED_CAST")
        val impressedProps = entries[0].payload["properties"] as? Map<String, Any?>
        assertNull(impressedProps?.get("element_id"))
    }

    @Test
    fun `batch threshold triggers immediate flush`() = testScope.runTest {
        val fakeSender = FakeSender { successResponse() }
        val service = createService(
            config = AnalyticsConfig(flushBatchSize = 2, flushIntervalMs = 10_000L),
            sender = fakeSender,
        )

        service.capture(DigiaExperienceEvent.Impressed, buildPayload("p1"))
        service.capture(DigiaExperienceEvent.Impressed, buildPayload("p2"))
        // second capture triggers batch flush (size >= flushBatchSize=2); run the dispatch coroutine
        advanceUntilIdle()

        assertEquals(0, service.queue.size())
        assertEquals(1, fakeSender.callCount)
    }

    @Test
    fun `timer fires after flushIntervalMs`() = testScope.runTest {
        val fakeSender = FakeSender { successResponse() }
        val service = createService(
            config = AnalyticsConfig(flushBatchSize = 10, flushIntervalMs = 50L),
            sender = fakeSender,
        )

        service.capture(DigiaExperienceEvent.Impressed, buildPayload("p1"))
        // capture is synchronous; timer scheduled at 50ms — do not call advanceUntilIdle here

        advanceTimeBy(200L)
        advanceUntilIdle()

        assertEquals(0, service.queue.size())
        assertEquals(1, fakeSender.callCount)
    }

    @Test
    fun `explicit flush dispatches pending events`() = testScope.runTest {
        val fakeSender = FakeSender { successResponse() }
        val service = createService(
            config = AnalyticsConfig(flushBatchSize = 10, flushIntervalMs = 10_000L),
            sender = fakeSender,
        )

        service.capture(DigiaExperienceEvent.Impressed, buildPayload("p1"))
        // capture is synchronous — event is immediately in queue
        assertEquals(1, service.queue.size())

        service.flush()
        advanceUntilIdle()

        assertEquals(0, service.queue.size())
        assertEquals(1, fakeSender.callCount)
    }

    @Test
    fun `5xx response retries and event survives until success`() = testScope.runTest {
        val fakeSender = FakeSender { callNum, _ ->
            if (callNum == 1) serverErrorResponse() else successResponse()
        }
        val service = createService(
            config = AnalyticsConfig(flushBatchSize = 10, flushIntervalMs = 10_000L),
            sender = fakeSender,
        )
        service.retryScheduleMs = listOf(10L, 20L)

        service.capture(DigiaExperienceEvent.Impressed, buildPayload("p1"))
        // capture is synchronous — do not advanceUntilIdle here (would advance clock to 10_000ms and fire timer)

        service.flush()
        runCurrent() // runs the flush dispatch at time=0 only; retry is scheduled in the future

        assertEquals(1, service.retryAttempt)
        assertEquals(1, service.queue.size())

        advanceTimeBy(100L)
        advanceUntilIdle() // retry fires → 200

        assertEquals(0, service.queue.size())
        assertEquals(0, service.retryAttempt)
        assertEquals(2, fakeSender.callCount)
    }

    @Test
    fun `lifecycle stop flushes pending events`() = testScope.runTest {
        val fakeSender = FakeSender { successResponse() }
        val service = createService(
            config = AnalyticsConfig(flushBatchSize = 10, flushIntervalMs = 10_000L),
            sender = fakeSender,
        )

        service.capture(DigiaExperienceEvent.Impressed, buildPayload("p1"))
        // capture is synchronous
        assertEquals(1, service.queue.size())

        service.onLifecycleStop()
        advanceUntilIdle()

        assertEquals(0, service.queue.size())
        assertEquals(1, fakeSender.callCount)
    }

    @Test
    fun `persisted queue flushes on next cold init`() = testScope.runTest {
        val store = InMemoryKeyValueStore()

        // First session — enqueue then simulate process death
        val service1 = createService(
            config = AnalyticsConfig(flushBatchSize = 10, flushIntervalMs = 10_000L),
            store = store,
        )
        service1.capture(DigiaExperienceEvent.Impressed, buildPayload("persisted"))
        // capture is synchronous — do not advanceUntilIdle (would fire timer and flush)
        assertEquals(1, service1.queue.size())
        service1.resetForTest()

        // Second session — re-init with same store, expects auto-flush via timer
        val fakeSender = FakeSender { successResponse() }
        createService(
            config = AnalyticsConfig(flushBatchSize = 10, flushIntervalMs = 50L),
            sender = fakeSender,
            store = store,
        )

        advanceTimeBy(200L)
        advanceUntilIdle()

        assertEquals(1, fakeSender.callCount)
        // queue belongs to service2 which shares the store
        assertEquals(0, AnalyticsQueue(store).size())
    }

    @Test
    fun `dismissed event queues but does not self-flush`() = testScope.runTest {
        val fakeSender = FakeSender { successResponse() }
        val service = createService(
            config = AnalyticsConfig(flushBatchSize = 100, flushIntervalMs = 10_000L),
            sender = fakeSender,
        )

        service.capture(DigiaExperienceEvent.Dismissed, buildPayload("p1"))
        // capture is synchronous — do not advanceUntilIdle (would fire timer)

        assertEquals(1, service.queue.size())
        assertEquals(0, fakeSender.callCount)
    }

    @Test
    fun `partial failure removes all batched events without retry`() = testScope.runTest {
        val fakeSender = FakeSender { _, _ ->
            // 207 partial failure — one accepted, one rejected
            SenderResult(207, """{"accepted":1,"rejected":1,"errors":[{"event_id":"any","reason":"invalid"}]}""")
        }
        val service = createService(
            config = AnalyticsConfig(flushBatchSize = 10, flushIntervalMs = 10_000L),
            sender = fakeSender,
        )

        service.capture(DigiaExperienceEvent.Impressed, buildPayload("p1"))
        service.capture(DigiaExperienceEvent.Clicked("cta"), buildPayload("p2"))
        // capture is synchronous

        service.flush()
        advanceUntilIdle()

        // Both events removed — partial failure is not retried
        assertEquals(0, service.queue.size())
        assertEquals(1, fakeSender.callCount)
    }
}

// ── Test doubles ─────────────────────────────────────────────────────────────

internal class InMemoryKeyValueStore : KeyValueStore {
    private val map = mutableMapOf<String, String>()
    override fun getString(key: String, default: String?) = map[key] ?: default
    override fun putString(key: String, value: String) { map[key] = value }
    override fun remove(key: String) { map.remove(key) }
}

internal class FakeSender(
    private val respond: (callNum: Int, body: String) -> SenderResult = { _, _ -> successResponse() },
) : AnalyticsSender {
    var callCount = 0

    // Convenience constructor for response-only lambdas
    constructor(respond: () -> SenderResult) : this({ _, _ -> respond() })

    override suspend fun post(
        url: String,
        jsonBody: String,
        headers: Map<String, String>,
    ): SenderResult = respond(++callCount, jsonBody)
}

private fun successResponse() =
    SenderResult(200, """{"accepted":1,"rejected":0,"errors":[]}""")

private fun serverErrorResponse() =
    SenderResult(500, """{"message":"internal server error"}""")

File: flutter/test/analytics_service_test.dart
import 'package:digia_engage/api/internal/event/engage_analytics_event.dart';
import 'package:digia_engage/api/models/analytics_config.dart';
import 'package:digia_engage/api/models/cep_trigger_payload.dart';
import 'package:digia_engage/api/models/digia_experience_event.dart';
import 'package:digia_engage/src/analytics/analytics_service.dart';
import 'package:digia_engage/src/preferences_store.dart';
import 'package:dio/dio.dart';
import 'package:flutter/widgets.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:package_info_plus/package_info_plus.dart';
import 'package:shared_preferences/shared_preferences.dart';

class _FakeDio implements Dio {
  _FakeDio(this.onPost);

  final Future<Response<Object?>> Function(
    String path, {
    Object? data,
    Map<String, dynamic>? queryParameters,
    Options? options,
    CancelToken? cancelToken,
    ProgressCallback? onSendProgress,
    ProgressCallback? onReceiveProgress,
  }) onPost;

  @override
  Future<Response<T>> post<T>(
    String path, {
    Object? data,
    Map<String, dynamic>? queryParameters,
    Options? options,
    CancelToken? cancelToken,
    ProgressCallback? onSendProgress,
    ProgressCallback? onReceiveProgress,
  }) {
    return onPost(
      path,
      data: data,
      queryParameters: queryParameters,
      options: options,
      cancelToken: cancelToken,
      onSendProgress: onSendProgress,
      onReceiveProgress: onReceiveProgress,
    ) as Future<Response<T>>;
  }

  @override
  dynamic noSuchMethod(Invocation invocation) => super.noSuchMethod(invocation);
}

void main() {
  TestWidgetsFlutterBinding.ensureInitialized();

  setUp(() async {
    SharedPreferences.setMockInitialValues({});
    await PreferencesStore.instance.initialize();
    PackageInfo.setMockInitialValues(
      appName: 'TestApp',
      packageName: 'com.test.app',
      version: '1.0.0',
      buildNumber: '1',
      buildSignature: '',
    );
    await DigiaAnalyticsService.instance.resetForTest();
  });

  Response<Object?> _makeResponse({
    required int statusCode,
    required Map<String, dynamic> data,
  }) {
    return Response<Object?>(
      requestOptions: RequestOptions(path: ''),
      statusCode: statusCode,
      data: data,
    );
  }

  Response<Object?> _successResponse() {
    return _makeResponse(
      statusCode: 200,
      data: {'accepted': 1, 'rejected': 0, 'errors': []},
    );
  }

  Response<Object?> _serverErrorResponse() {
    return _makeResponse(
      statusCode: 500,
      data: {'message': 'internal server error'},
    );
  }

  Response<Object?> _partialFailureResponse(String failedEventId) {
    return _makeResponse(
      statusCode: 200,
      data: {
        'accepted': 1,
        'rejected': 1,
        'errors': [
          {'event_id': failedEventId, 'reason': 'invalid campaign_type'}
        ],
      },
    );
  }

  test('initializes analytics and persists anonymous id', () async {
    await DigiaAnalyticsService.instance.initialize(
      const DigiaAnalyticsConfig(),
      'digia_test',
    );

    final anonymousId = DigiaAnalyticsService.instance.getAnonymousId();
    expect(anonymousId, isNotEmpty);

    final secondId = DigiaAnalyticsService.instance.getAnonymousId();
    expect(secondId, equals(anonymousId));
  });

  test('setUserId and clearUserId persist and reset session', () async {
    await DigiaAnalyticsService.instance.initialize(
      const DigiaAnalyticsConfig(),
      'digia_test',
    );

    await DigiaAnalyticsService.instance.setUserId('user-123');
    expect(DigiaAnalyticsService.instance.userId, equals('user-123'));

    final sessionBefore = DigiaAnalyticsService.instance.sessionId;
    await DigiaAnalyticsService.instance.clearUserId();
    expect(DigiaAnalyticsService.instance.userId, isNull);
    expect(DigiaAnalyticsService.instance.sessionId, isNotEmpty);
    expect(
        DigiaAnalyticsService.instance.sessionId, isNot(equals(sessionBefore)));
  });

  test('queue drops oldest events when capacity is exceeded', () async {
    await DigiaAnalyticsService.instance.initialize(
      const DigiaAnalyticsConfig(queueMaxEvents: 3),
      'digia_test',
    );

    for (var i = 0; i < 5; i++) {
      final payload = _buildPayload('event-$i');
      await DigiaAnalyticsService.instance.capture(
        const CarouselViewed(),
        payload,
      );
      await Future<void>.delayed(const Duration(milliseconds: 1));
    }

    final queueLength = await DigiaAnalyticsService.instance.getQueueLength();
    expect(queueLength, equals(3));

    final entries = await DigiaAnalyticsService.instance.debugQueueEntries();
    expect(entries.length, equals(3));

    final createdAtValues = entries
        .map((entry) => entry['created_at'] as int)
        .toList(growable: false);
    final ordered = [...createdAtValues]..sort();
    expect(createdAtValues, equals(ordered));
  });

  test('builds event payload with static context and identity fields',
      () async {
    await DigiaAnalyticsService.instance.initialize(
      const DigiaAnalyticsConfig(),
      'digia_test',
    );

    final payload = _buildPayload('payload-1');
    await DigiaAnalyticsService.instance.capture(
      const CarouselViewed(),
      payload,
      campaignId: 'example-campaign',
    );

    final entries = await DigiaAnalyticsService.instance.debugQueueEntries();
    expect(entries, isNotEmpty);
    final event = entries.first['payload'] as Map<String, dynamic>;

    expect(event['event_name'], equals('Digia Experience Viewed'));
    expect(event['campaign_id'], equals('example-campaign'));
    expect((event['properties'] as Map)['sdk_version'], isNotNull);
    expect(event['anonymous_id'], isNotEmpty);
    expect(event['session_id'], isNotEmpty);
  });

  test('experience event names and flush behavior live in the event model', () {
    expect(const ExperienceImpressed().analyticsEventName,
        equals('Digia Experience Viewed'));
    expect(const ExperienceClicked(elementId: 'cta').analyticsEventName,
        equals('Digia Experience Clicked'));
    expect(const ExperienceDismissed().analyticsEventName,
        equals('Digia Experience Dismissed'));
    expect(const ExperienceCompleted().analyticsEventName,
        equals('Digia Experience Completed'));

    expect(const ExperienceImpressed().flushOnCapture, isFalse);
    expect(const ExperienceClicked().flushOnCapture, isFalse);
    expect(const ExperienceDismissed().flushOnCapture, isTrue);
    expect(const ExperienceCompleted().flushOnCapture, isTrue);
  });

  test('flushes immediately when queue reaches flushBatchSize', () async {
    var postCount = 0;
    final fakeDio = _FakeDio(
      (path,
          {data,
          queryParameters,
          options,
          cancelToken,
          onSendProgress,
          onReceiveProgress}) async {
        postCount += 1;
        return _makeResponse(
          statusCode: 200,
          data: {'accepted': 2, 'rejected': 0, 'errors': []},
        );
      },
    );

    await DigiaAnalyticsService.instance.initialize(
      const DigiaAnalyticsConfig(flushBatchSize: 2, flushIntervalMs: 10000),
      'digia_test',
    );
    DigiaAnalyticsService.instance.dioFactory = () => fakeDio;

    await DigiaAnalyticsService.instance.capture(
      const CarouselViewed(),
      _buildPayload('payload-1'),
    );
    await DigiaAnalyticsService.instance.capture(
      const CarouselViewed(),
      _buildPayload('payload-2'),
    );

    expect(await DigiaAnalyticsService.instance.getQueueLength(), equals(0));
    expect(postCount, equals(1));
  });

  test('schedules flush after flushIntervalMs when threshold is not reached',
      () async {
    var postCount = 0;
    final fakeDio = _FakeDio(
      (path,
          {data,
          queryParameters,
          options,
          cancelToken,
          onSendProgress,
          onReceiveProgress}) async {
        postCount += 1;
        return _successResponse();
      },
    );

    await DigiaAnalyticsService.instance.initialize(
      const DigiaAnalyticsConfig(flushBatchSize: 10, flushIntervalMs: 50),
      'digia_test',
    );
    DigiaAnalyticsService.instance.dioFactory = () => fakeDio;

    await DigiaAnalyticsService.instance.capture(
      const CarouselViewed(),
      _buildPayload('payload-1'),
    );

    await Future<void>.delayed(const Duration(milliseconds: 200));

    expect(await DigiaAnalyticsService.instance.getQueueLength(), equals(0));
    expect(postCount, equals(1));
  });

  test('explicit flush() dispatches pending events', () async {
    var postCount = 0;
    final fakeDio = _FakeDio(
      (path,
          {data,
          queryParameters,
          options,
          cancelToken,
          onSendProgress,
          onReceiveProgress}) async {
        postCount += 1;
        return _successResponse();
      },
    );

    await DigiaAnalyticsService.instance.initialize(
      const DigiaAnalyticsConfig(flushBatchSize: 10, flushIntervalMs: 10000),
      'digia_test',
    );
    DigiaAnalyticsService.instance.dioFactory = () => fakeDio;

    await DigiaAnalyticsService.instance.capture(
      const CarouselViewed(),
      _buildPayload('payload-1'),
    );
    expect(await DigiaAnalyticsService.instance.getQueueLength(), equals(1));

    await DigiaAnalyticsService.instance.flush();

    expect(await DigiaAnalyticsService.instance.getQueueLength(), equals(0));
    expect(postCount, equals(1));
  });

  test('retries transient failures and preserves the event until success',
      () async {
    var callCount = 0;
    final fakeDio = _FakeDio(
      (path,
          {data,
          queryParameters,
          options,
          cancelToken,
          onSendProgress,
          onReceiveProgress}) async {
        callCount += 1;
        if (callCount == 1) {
          return _serverErrorResponse();
        }
        return _successResponse();
      },
    );

    await DigiaAnalyticsService.instance.initialize(
      const DigiaAnalyticsConfig(flushBatchSize: 10, flushIntervalMs: 10000),
      'digia_test',
    );
    DigiaAnalyticsService.instance.dioFactory = () => fakeDio;
    DigiaAnalyticsService.instance.retryScheduleMs = [10, 20];

    await DigiaAnalyticsService.instance.capture(
      const CarouselViewed(),
      _buildPayload('payload-1'),
    );

    await DigiaAnalyticsService.instance.flush();
    expect(DigiaAnalyticsService.instance.retryAttempt, equals(1));
    expect(await DigiaAnalyticsService.instance.getQueueLength(), equals(1));

    await Future<void>.delayed(const Duration(milliseconds: 100));

    expect(await DigiaAnalyticsService.instance.getQueueLength(), equals(0));
    expect(DigiaAnalyticsService.instance.retryAttempt, equals(0));
    expect(callCount, equals(2));
  });

  test('appLifecycleChanged paused flushes pending events', () async {
    var postCount = 0;
    final fakeDio = _FakeDio(
      (path,
          {data,
          queryParameters,
          options,
          cancelToken,
          onSendProgress,
          onReceiveProgress}) async {
        postCount += 1;
        return _successResponse();
      },
    );

    await DigiaAnalyticsService.instance.initialize(
      const DigiaAnalyticsConfig(flushBatchSize: 10, flushIntervalMs: 10000),
      'digia_test',
    );
    DigiaAnalyticsService.instance.dioFactory = () => fakeDio;

    await DigiaAnalyticsService.instance.capture(
      const CarouselViewed(),
      _buildPayload('payload-1'),
    );
    expect(await DigiaAnalyticsService.instance.getQueueLength(), equals(1));

    await DigiaAnalyticsService.instance
        .appLifecycleChanged(AppLifecycleState.paused);

    expect(await DigiaAnalyticsService.instance.getQueueLength(), equals(0));
    expect(postCount, equals(1));
  });

  test('initialize schedules flush timer when persisted queue exists',
      () async {
    var postCount = 0;
    final fakeDio = _FakeDio(
      (path,
          {data,
          queryParameters,
          options,
          cancelToken,
          onSendProgress,
          onReceiveProgress}) async {
        postCount += 1;
        return _successResponse();
      },
    );

    await DigiaAnalyticsService.instance.initialize(
      const DigiaAnalyticsConfig(flushBatchSize: 10, flushIntervalMs: 50),
      'digia_test',
    );
    DigiaAnalyticsService.instance.dioFactory = () => fakeDio;

    await DigiaAnalyticsService.instance.capture(
      const CarouselViewed(),
      _buildPayload('payload-opened'),
    );
    expect(await DigiaAnalyticsService.instance.getQueueLength(), equals(1));

    await DigiaAnalyticsService.instance.resetForTest();

    await DigiaAnalyticsService.instance.initialize(
      const DigiaAnalyticsConfig(flushBatchSize: 10, flushIntervalMs: 50),
      'digia_test',
    );
    DigiaAnalyticsService.instance.dioFactory = () => fakeDio;

    await Future<void>.delayed(const Duration(milliseconds: 150));

    expect(await DigiaAnalyticsService.instance.getQueueLength(), equals(0));
    expect(postCount, equals(1));
  });

  test('dismissed event does not flush inside analytics service', () async {
    var postCount = 0;
    final fakeDio = _FakeDio(
      (path,
          {data,
          queryParameters,
          options,
          cancelToken,
          onSendProgress,
          onReceiveProgress}) async {
        postCount += 1;
        return _successResponse();
      },
    );

    await DigiaAnalyticsService.instance.initialize(
      const DigiaAnalyticsConfig(flushBatchSize: 100, flushIntervalMs: 10000),
      'digia_test',
    );
    DigiaAnalyticsService.instance.dioFactory = () => fakeDio;

    await DigiaAnalyticsService.instance.capture(
      const SurveyDismissed(),
      _buildPayload('payload-dismissed'),
    );

    expect(await DigiaAnalyticsService.instance.getQueueLength(), equals(1));
    expect(postCount, equals(0));
  });

  test('partial failure response removes accepted and rejected events',
      () async {
    var callCount = 0;
    String failedEventId = '';
    final fakeDio = _FakeDio(
      (path,
          {data,
          queryParameters,
          options,
          cancelToken,
          onSendProgress,
          onReceiveProgress}) async {
        callCount += 1;
        final body = data as Map<String, dynamic>;
        final eventIds = (body['events'] as List)
            .map((event) => event['event_id'] as String)
            .toList();
        failedEventId = eventIds.first;
        return _partialFailureResponse(failedEventId);
      },
    );

    await DigiaAnalyticsService.instance.initialize(
      const DigiaAnalyticsConfig(flushBatchSize: 10, flushIntervalMs: 10000),
      'digia_test',
    );
    DigiaAnalyticsService.instance.dioFactory = () => fakeDio;

    await DigiaAnalyticsService.instance.capture(
      const CarouselViewed(),
      _buildPayload('payload-1'),
    );
    await DigiaAnalyticsService.instance.capture(
      const CarouselClicked(elementId: 'cta'),
      _buildPayload('payload-2'),
      campaignId: 'digia-campaign-123',
    );

    await DigiaAnalyticsService.instance.flush();

    expect(await DigiaAnalyticsService.instance.getQueueLength(), equals(0));
    expect(callCount, equals(1));
  });
}

CEPTriggerPayload _buildPayload(String eventKey) {
  return CEPTriggerPayload(
    cepCampaignId: 'example-campaign',
    cepMetadata: {
      'key': eventKey,
      'type': 'guide',
      'displayStyle': 'modal',
      'anchorKey': 'button-1',
      'slotKey': 'slot-1',
    },
    campaignKey: eventKey,
    variables: {'user_name': 'test-user'},
  );
}
